File: java/src/main/java/org/anhonesteffort/algosndatas/Queue.java
/*
 * Copyright (C) 2016 An Honest Effort LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.anhonesteffort.algosndatas;

/*
  QUEUES ARE FIRST IN FIRST OUT.
 */
public class Queue<T> {

  private Item<T> head = null;
  private Item<T> tail = null;

  public void enqueue(T value) {
    if (head == null) {
      head = new Item<>(value);
      tail = head;
    } else {
      head.left = new Item<>(value);
      head      = head.left;
    }
  }

  public T peek() {
    return (tail != null) ? tail.value : null;
  }

  public T dequeue() {
    if (tail != null && tail.left != null) {
      T value = tail.value;
      tail = tail.left;
      return value;
    } else if (tail != null) {
      T value = tail.value;
      head = tail = null;
      return value;
    } else {
      return null;
    }
  }

  private class Item<V extends T> {
    private final V value;
    private Item<V> left;

    public Item(V value) {
      this.value = value;
    }
  }

}

File: js/test/queue.js
var test = require('tape');
/*
 * QUEUES ARE FIRST IN FIRST OUT.
 */


var Item = function(value) {
  this.value = value;
}

var Queue = function() {
  this.head = undefined;
  this.tail = undefined;
}

Queue.prototype.enqueue = function(value) {
  if (this.head === undefined) {
    this.head = new Item(value);
    this.tail = this.head;
  } else {
    this.head.left = new Item(value);
    this.head      = this.head.left;
  }
}

Queue.prototype.peek = function() {
  return (this.tail !== undefined) ? this.tail.value : undefined;
}

Queue.prototype.dequeue = function() {
  if (this.tail !== undefined && this.tail.left !== undefined) {
    var value = this.tail.value;
    this.tail = this.tail.left;
    return value;
  } else if (this.tail !== undefined) {
    var value = this.tail.value;
    this.head = this.tail = undefined;
    return value;
  } else {
    return undefined;
  }
}


test('testQueue', function (t) {
  t.plan(14);

  var queue = new Queue();
  t.equal(queue.peek(),    undefined);
  t.equal(queue.dequeue(), undefined);

  queue.enqueue(1);
  queue.enqueue(2);
  queue.enqueue(3);

  t.equal(queue.peek(),    1);
  t.equal(queue.dequeue(), 1);

  queue.enqueue(4);

  t.equal(queue.peek(),    2);
  t.equal(queue.dequeue(), 2);
  t.equal(queue.peek(),    3);
  t.equal(queue.dequeue(), 3);
  t.equal(queue.peek(),    4);
  t.equal(queue.dequeue(), 4);

  t.equal(queue.peek(),    undefined);
  t.equal(queue.dequeue(), undefined);

  queue.enqueue(5);

  t.equal(queue.peek(),    5);
  t.equal(queue.dequeue(), 5);
});
